src/openclsim/critical_path/dependencies_from_simpy_step.py

```diff
@@ -1,40 +1,245 @@
+
+        # get recorded activities and convert times to floats (seconds since Jan 1970)
+        recorded_activities_df = self.recorded_activities_df.copy()
+        recorded_activities_df.start_time = round(
+            recorded_activities_df.start_time.astype("int64") / 10**9, 4
+        )
+        recorded_activities_df.end_time = round(
+            recorded_activities_df.end_time.astype("int64") / 10**9, 4
+        )
+
+        # rename the dependencies from dependencies with e_id to dependencies with cp_activity_id
+        self.dependency_list = [
+            (
+                self._find_cp_act(dependency[0], recorded_activities_df),
+                self._find_cp_act(dependency[1], recorded_activities_df),
+            )
+            for dependency in DEPENDENCIES_OPENCLSIM
+        ]
+
+    def _find_cp_act(self, e_id, recorded_activities_df):
+        """
+        Get cp activity ID given a time-window and an activity ID.
+
+        Parameters
+        ----------
+        e_id : int
+            Execution id from simpy.
+        recorded_activities_df : pd.DataFrame
+            From self.get_recorded_activity_df().
+        """
+        activity_id = self.step_logging_dataframe.loc[e_id, "event_object"].value
+        end_time = round(self.step_logging_dataframe.loc[e_id, "t1"], 4)
+        matching_ids = recorded_activities_df.loc[
+            (
+                (recorded_activities_df.ActivityID == activity_id)
+                & (recorded_activities_df.end_time == end_time)
+            ),
+            "cp_activity_id",
+        ]
+        if len(set(matching_ids)) == 1:
+            cp_activity_id = matching_ids.iloc[0]
+        else:
+            raise ValueError(
+                f"No recorded activity found for"
+                f" activityID {activity_id} at (end)time {end_time}"
+            )
+        return cp_activity_id
+
+
+class AlteredStepEnv(simpy.Environment):
+    """
+    Class is child of simpy.Environment and passes on all arguments on initialization.
+    The 'step' method is overwritten (or 'monkey-patched') in order to log some data of
+    simulation into self.data_step and self.data_cause_effect. The former saves some metadata
+    of the Event such as e_id (execution ID), simulation time, prio and event type (list of tuples).
+    The latter saves which e_id scheduled another e_id and is hence a list of cause-effect tuples.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialization."""
+        super().__init__(*args, **kwargs)
+        self.data_cause_effect = []
+        self.data_step = []
+
+    def step(self):
+        """
+        The 'step' method is overwritten (or 'monkey-patched') in order to log some data of
+        simulation into self.data_step and self.data_cause_effect.
+        """
+        time_start = copy.deepcopy(self.now)
+        if len(self._queue):
+            _, prio, e_id, event = self._queue[0]
+            old_e_ids = set([t[2] for t in self._queue])
```

Comment:
maybe rename to ``e_ids_before`` and ``e_ids_after``?

Reply:
agree